storage/remote/queue_manager.go

```diff
@@ -467,21 +475,42 @@ func (s *shards) runShard(i int) {
 }
 
 func (s *shards) sendSamples(samples model.Samples) {
-	// Samples are sent to the remote storage on a best-effort basis. If a
-	// sample isn't sent correctly the first time, it's simply dropped on the
-	// floor.
 	begin := time.Now()
-	err := s.qm.cfg.Client.Store(samples)
+	s.sendSamplesWithBackoff(samples)
 	duration := time.Since(begin)
 
-	if err != nil {
-		log.Warnf("error sending %d samples to remote storage: %s", len(samples), err)
-		failedSamplesTotal.WithLabelValues(s.qm.queueName).Add(float64(len(samples)))
-	} else {
-		sentSamplesTotal.WithLabelValues(s.qm.queueName).Add(float64(len(samples)))
-	}
-	sentBatchDuration.WithLabelValues(s.qm.queueName).Observe(duration.Seconds())
-
+	// These counters are used to caclulate the dynamic sharding, as as such
+	// should be mainted irrespective of success or failure.
 	s.qm.samplesOut.incr(int64(len(samples)))
 	s.qm.samplesOutDuration.incr(int64(duration))
 }
+
+// sendSamples to the remote storage with backoff for recoverable errors.
+func (s *shards) sendSamplesWithBackoff(samples model.Samples) {
+	backoff := s.qm.cfg.MinBackoff
+	for retries := s.qm.cfg.MaxRetries; retries > 0; retries-- {
+		begin := time.Now()
+		err := s.qm.client.Store(samples)
+		duration := time.Since(begin)
```

Comment:
Since it's only used in one place now, I'd get rid of this var and just do `Observe(time.Since(begin))`.